### ISSUE TITLE
issue with docker setup : xmpp.meet.jitsi could not be resolved, but internal ping does work

### ISSUE THREAD

**Issue by squalou** [jitsi/docker-jitsi-meet#1813]
Hi,

I'm not a complete newbie in docker world but I'm facing something I really don't get. I'ts more my networking abilities that must be at fault.

I'm trying to setup jitsi on AWS/ECS, I found plenty documentation (yours being very good to begin with, so thank you for that !).
Configuration is close to what is given in docker-compose so it's not that hard to do.

I'm not yet completely done with UDP and other things, but for now, 4 containers are starting
- web
- prosody
- jicofo
- jvb

and I can reach my custom PUBLIC_URL over https, but THEN, when trying to start a meet, I get a "you've been disconnected" error.

In my browser console I can see : NS_ERROR_WEBSOCKET_CONNECTION_REFUSED

I received a 502 error from wss://<my-public-url>/xmpp-websocket?room=wtf

And most important : docker logs from the web container shows

```
[error] 281#281: *14 xmpp.meet.jitsi could not be resolved (3: Host not found), client: XX.XX.XX.XX, server: _, request: "GET /xmpp-websocket?room=wtf HTTP/1.1", host:  "<my-public-url>"
```

That's where I don't get it. for the records, in ECS vocabulary :

- I've set up a "hostname" on prosody container with "xmpp.meet.jitsi" as value
- I've set up other containers to link to prosody container with this name

As a result, If I `docker exe -ti  xxxxxxx bash`  ....into `web` container : `xmpp.meet.jitsi` can be reached from internal docker network

```
root@171ad8a0ddf8:/# ping xmpp.meet.jitsi
PING xmpp.meet.jitsi (172.17.0.7) 56(84) bytes of data.
64 bytes from xmpp.meet.jitsi (172.17.0.7): icmp_seq=1 ttl=127 time=0.097 ms
64 bytes from xmpp.meet.jitsi (172.17.0.7): icmp_seq=2 ttl=127 time=0.050 ms
64 bytes from xmpp.meet.jitsi (172.17.0.7): icmp_seq=3 ttl=127 time=0.038 ms
64 bytes from xmpp.meet.jitsi (172.17.0.7): icmp_seq=4 ttl=127 time=0.029 ms
64 bytes from xmpp.meet.jitsi (172.17.0.7): icmp_seq=5 ttl=127 time=0.050 ms

```

so there's something I'm missing, probably around name resolution from web application ? 

Any clue will be welcome.

thank you !

**Comment by saghul** [comment]
Perhaps the resolver that nginx is using? what do you have in /etc/resolv.conf ?

**Comment by squalou** [comment]
Inside web container you mean ?

```
nameserver 10.19.0.2
search eu-west-3.compute.internal
```

indeed this cannot resolve xmpp.meet.jitsi, for sure, but then how "ping" can do, that's a mystery.

Also, using `dig` also fails (but that makes sens for internal docker network)

Should it be another configuration ? I have no idea how this one arrives here, aws magic + resolveconf magic probably

**Comment by squalou** [comment]
Additionally : would there be a way to override "xmpp.meet.jitsi", so that I could use a custom named declared in a DNS ?

I came accross XMPP_SERVER variable (and XMPP_DOMAIN) but could not figure out their use.

**Comment by saghul** [comment]
I don't have experience with ECS, maybe this helps? https://aws.amazon.com/blogs/opensource/getting-started-with-jitsi-an-open-source-web-conferencing-solution/

**Comment by squalou** [comment]
thank you ... unfortunately that's the documentation I already used ;)

Your idea that it may be related to nginx is not bad, there's probably something there. I don't really get the high-level architecture of jitsi : is "xmpp.meet.jitsi" supposed to be a customisable (and eventuually) public name ?
Or is it meant to be hardcoded ?

**Comment by saghul** [comment]
You can change it, but we default to a non-routable domain, because Docker itself is supposed to know how to route that.

**Comment by squalou** [comment]
ok,  that was my initial understanding.

In the meantime if I was looking for a workaround, how would I do this ?
I mean, 
I see this : 

```
$ rg xmpp.meet.jitsi

prosody/rootfs/defaults/conf.d/visitors.cfg.lua
36:{{ $XMPP_SERVER := .Env.XMPP_SERVER | default "xmpp.meet.jitsi" -}}

docker-compose.yml
312:                    - ${XMPP_SERVER:-xmpp.meet.jitsi}

jibri/rootfs/defaults/xmpp.conf
14:{{ $XMPP_SERVER := .Env.XMPP_SERVER | default "xmpp.meet.jitsi" -}}

jvb/rootfs/defaults/jvb.conf
26:{{ $XMPP_SERVER := .Env.XMPP_SERVER | default "xmpp.meet.jitsi" -}}

jvb/rootfs/etc/cont-init.d/10-config
14:[ -z "${XMPP_SERVER}" ] && export XMPP_SERVER=xmpp.meet.jitsi

jigasi/rootfs/defaults/sip-communicator.properties
10:{{ $XMPP_SERVER := .Env.XMPP_SERVER | default "xmpp.meet.jitsi" -}}

jicofo/rootfs/defaults/jicofo.conf
37:{{ $XMPP_SERVER := .Env.XMPP_SERVER | default "xmpp.meet.jitsi" -}}

```

but having defined XMPP_SERVER to a custom environment variable seems to have no effect.

I do the equivalent of `docker run -e XMPP_SERVER=my_value` but I'm not sure it is the same as having a `.env` file given the `.Env.XMPP_SERVER` syntax.

**Comment by saghul** [comment]
I'd need to check, a bug might have crept in, since that is not something that is, in general, necessary.

Can you check if 127.0.0.11 is able to resolve the IP?

**Comment by squalou** [comment]
Sure,

```
root@171ad8a0ddf8:/# dig 127.0.0.11

; <<>> DiG 9.18.24-1-Debian <<>> 127.0.0.11
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 55317
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;127.0.0.11.			IN	A

;; ANSWER SECTION:
127.0.0.11.		0	IN	A	127.0.0.11
```

ping works too.


Note that I have a (hopefully) interesting thing : `/etc/hosts` contains `xmpp.meet.jitsi`

So, it could explain why ping works,
but I dont get why name resolution would fail if things are in /etc/hosts.

Here it is :

```
root@171ad8a0ddf8:/# cat /etc/hosts
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.17.0.7	xmpp.meet.jitsi xmpp.meet.jitsi ecs-jitsi-8-prosody-a8a49dbabcf4aaf12100
172.17.0.10	171ad8a0ddf8

```

**Comment by squalou** [comment]
maybe related to this sort of thing ?

https://stackoverflow.com/questions/29980884/proxy-pass-does-not-resolve-dns-using-etc-hosts

In case it's nginx that is in use for the resolution ?

**Comment by saghul** [comment]
That resolver is taken from /etc/resolv.conf by our startup code, if you don't set it:

https://github.com/jitsi/docker-jitsi-meet/blob/cacee4ecc817361905f04930f24ff2273502972a/web/rootfs/etc/cont-init.d/10-config#L67

The IP I asked before, is it able to resolve the XMPP server domain?

**Comment by squalou** [comment]
Thx, I feraed it was somehow a kind of AWS magically set variable somewhere.

I tried to add this in nginx.conf  instead of existing resolver line : 

```
resolver 127.0.0.11 ipv6=off;
```

With or wihout the ipv6 off I unfortunately get this now :

```
error] 1001#1001: recv() failed (111: Connection refused) while resolving, resolver: 127.0.0.11:53
```

I fear this internal resolver is not present, or not at this address, no idea how to make sure. I'll probably end up calling aws support just in case. (there are plenty of people struggling with it from my internet searchs... but no solution so far)

**Comment by squalou** [comment]
OMFG!

Here' what I did - not sure if it's a solution that can be used in real life.

I installed dnsmasq with this config : 

```
#Define Port
port=53
domain-needed
bogus-priv

#dont use hosts nameservers
no-resolv

strict-order
#Use the default nameservers
server=10.19.0.2
expand-hosts
```

I have no idea is this configuration makes sense, is missing things, anyway that's for a quick test . Yet another tunnel to dig.


Then `/etc/init.d/dnsmasq start`

And in `nginx.conf` I set `resolver 127.0.0.1;`

and apparently things now work.

That nails the issue as being AWS ECS related around name resolution.

Now, I wonder if I should go this way. That would mean

- custom images with dnsmasq in them (probably not a big deal)
- custom entrypoint.sh (probably harder but doable ?)
- override nginx conf (doable)


Thank for your help @saghul !

I'll keep this issue open for a time if you don't mind, and will add final words for whoever might come by one day.

**Comment by squalou** [comment]
Finally here's the 'quick' fix I did, the custom script will have to be improved but it gives an idea.


## 1.Dockerfile

```
FROM docker.io/jitsi/web:stable-9457-2
RUN apt-get update && apt-get install -y dnsmasq 
ADD 99-custom /etc/cont-init.d/
```

## 2. File 99-custom

```
#!/bin/bash
# this part comes from 10-config to find initial resolver IP
IP_LIST=""

# Parse IPs in /etc/resolv.conf, taking into account IPv6 addresses need to be
# enclosed in square brackets for the Nginx config file.
while read -r line; do
    if [[ $line =~ ^nameserver.* ]]; then
        IP=$(echo $line | cut -d" " -f2)
        COLONS=$(echo $IP | tr -dc ":" | awk '{ print length '})
        if [[ $COLONS -ge 2 ]]; then
            IP="[$IP]"
        fi
        if [[ ! "$IP_LIST" = "" ]]; then
            IP_LIST+=" "
        fi
        IP_LIST+="$IP"
    fi
done < <(cat /etc/resolv.conf)

DNSMASQ_RESOLV_IP=$(echo $IP_LIST | cut -d ' ' -f1)

cat << EOF > /etc/dnsmasq.d/custom
#Define Port
port=53
domain-needed
bogus-priv

#dont use hosts nameservers
no-resolv

strict-order
#Use the default nameservers
server=$DNSMASQ_RESOLV_IP
expand-hosts

EOF

cat /etc/dnsmasq.d/boxtal

/etc/init.d/dnsmasq start
/etc/init.d/dnsmasq status
exit 0

```

## 3.  Nginx conf

Start the container with  `-e NGINX_RESOLVER=127.0.0.1` to set nginx resolver.

**Comment by saghul** [comment]
Glad you solved it! It's really weird, because the resolver where dnsmasq is sending the queries is the same as nginx was...

**Comment by squalou** [comment]
yeah I thin the key element here is `expand-hosts`  option in dnsmasq.

My understanding is that with this option, dnsmasq *uses* /etc/hosts, whereas other resolver do *not*.

And Nginx simply relies on whataver dns resolver is in charge.

My last guess : probably docker-compose also does some more things, or plain docker, in order to achieve this but AWS ECS does not.

**Comment by saghul** [comment]
You are 100% correct indeed!

**Comment by gitbhavesh07** [comment]
@squalou, will you be able to share the task defn, any CFN template, or terraform scripts?

**Comment by AndrewV6** [comment]
A bit simplier solution for customized XMPP_SERVER
```yaml
    # XMPP server
    prosody:
        image:
        <...>
        networks:
            meet.jitsi:
                aliases:
                    - xmpp.meet.jitsi
                    - ${XMPP_SERVER:-xmpp.meet.jitsi}
```

**Comment by ashleyghooper** [comment]
I had a case today where I was getting the jitsi docker compose stack set up and didn't have all of the config right, so on my first attempt to start things up, one or more of the containers had errors and something prevented the name `xmpp.meet.jitsi` from being resolvable in that container, while it would resolve okay from the other containers. I don't remember exactly the sequence of events leading up to that - possibly that container got restarted, or one or more the others did.

Anyway, I found that by restarting the entire stack with `docker compose down`, then `docker compose up -d` the name resolution issue was fixed and everything started working.
